### PR TITLE
Fix dropping sharding on 0-sized boolean masks (#35273)

### LINKED ISSUE #35273: Inconsistent value dependent device for boolean masked array
### Description

On a colab notebook with a CUDA device I observe the following:

```python
import numpy as np
import jax.numpy as jnp
import jax

cpu_device = jax.devices("cpu")[0]
cpu_array = jnp.asarray([1], device=cpu_device)
masked = cpu_array[jnp.asarray([True], device=cpu_device)]
masked.shape, masked.device
```
```
((1,), CpuDevice(id=0))
```

Which is what I would expect.

However, when the masking result in a 0-sized dimension, I get a result on the CUDA device instead of the CPU device I would have expected given the explicit device selection the above code:

```python
masked = cpu_array[jnp.asarray([False], device=cpu_device)]
masked.shape, masked.device
```
```
((0,), CudaDevice(id=0))
```

### System info (python version, jaxlib version, accelerator, etc.)

```


jax:    0.7.2
jaxlib: 0.7.2
numpy:  2.0.2
python: 3.12.12 (main, Oct 10 2025, 08:52:57) [GCC 11.4.0]
device info: Tesla T4-1, 1 local devices"
process_count: 1
platform: uname_result(system='Linux', node='966a76ee8f69', release='6.6.105+', version='#1 SMP Thu Oct  2 10:42:05 UTC 2025', machine='x86_64')

$ nvidia-smi
Fri Feb 20 16:12:03 2026       
+-----------------------------------------------------------------------------------------+
| NVIDIA-SMI 580.82.07              Driver Version: 580.82.07      CUDA Version: 13.0     |
+-----------------------------------------+------------------------+----------------------+
| GPU  Name                 Persistence-M | Bus-Id          Disp.A | Volatile Uncorr. ECC |
| Fan  Temp   Perf          Pwr:Usage/Cap |           Memory-Usage | GPU-Util  Compute M. |
|                                         |                        |               MIG M. |
|=========================================+========================+======================|
|   0  Tesla T4                       Off |   00000000:00:04.0 Off |                    0 |
| N/A   71C    P0             32W /   70W |   11305MiB /  15360MiB |      0%      Default |
|                                         |                        |                  N/A |
+-----------------------------------------+------------------------+----------------------+

+-----------------------------------------------------------------------------------------+
| Processes:                                                                              |
|  GPU   GI   CI              PID   Type   Process name                        GPU Memory |
|        ID   ID                                                               Usage      |
|=========================================================================================|
|    0   N/A  N/A            2061      C   /usr/bin/python3                      11302MiB |
+-----------------------------------------------------------------------------------------+
```

### PR DESCRIPTION
**Pull request by Ashutosh0x**
Fixes jax-ml#35273

This PR replaces earlier PR attempts (#35292, #35293).

### Root Cause
When indexing arrays with boolean masks that evaluate to size 0, JAX propagates an abstract `NamedSharding` with an empty mesh during the lowering process. Inside `lax.full_like` (which powers the zero-filling for the unused mask entries), this abstract sharding overrides the physical device ID of the source array. Thus, memory gets allocated on the default device instead of the intended one.

### Fix
Instead of intercepting this at the top `indexing.py` layer (as done in older, closed PRs), this fix makes `lax.full_like` architecturally immune to empty-mesh shardings. 
It checks if `type(sharding) is NamedSharding and sharding.mesh.empty` and if so, treats the sharding as unspecified. This correctly trips the fallback logic, propagating the true concrete `sharding` (with device IDs intact) from the original array.

### CODE REVIEW

**Comment by Ashutosh0x** [comment]
I've addressed the implementation feedback (Tracer support and robust attribute checks) in a new clean PR: #35297
